src/core/query.ts

```diff
@@ -378,8 +378,8 @@ export class Query<
     }
 
     // Create query function context
-    const queryKey = ensureArray(this.queryKey)
-    const queryFnContext: QueryFunctionContext<unknown[]> = {
+    const queryKey = (ensureArray(this.queryKey) as unknown) as TQueryKey
+    const queryFnContext: QueryFunctionContext<TQueryKey> = {
```

Comment:
Could not find good alternatives to `as unknown`


```ts
    const queryKey = ensureArray(this.queryKey)
    const queryFnContext: QueryFunctionContext<TQueryKey> = {
      queryKey, // Type 'unknown[]' is not assignable to type 'TQueryKey'.
      pageParam: undefined,
    }
```


```ts
const queryKey: TQueryKey = ensureArray(this.queryKey) // Type 'unknown[]' is not assignable to type 'TQueryKey'.
```

Reply:
The problem seems to be that `ensureArray` narrows the type to `Array`, but `QueryFunctionContext` still wants us to give it a `TQueryKey`. `TQueryKey` can, on type level, also be e.g. a `string`, but it cannot be that after `ensureArray`.

I would also question why we do the `ensureArray` here in the first place, maybe @boschni knows? Because it means that the `queryKey` we get into our context is actually not the same as the queryKey we pass into the queryFn. example:

```
  const queryInfo= useQuery(
    "test",
    async (context) => {
      console.log(context.queryKey.toUpperCase());
      return 5;
    }
  )
```

here, context.queryKey will be an Array: `["test"]` at runtime, but if we do the type assertion like we do it here, it will be of type `"test"`. This means that typescript is fine calling `toUpperCase` on the queryKey, but it will fail at runtime.

---

so maybe we should just remove the `ensureArray` call? Then we'll have no problems with the types, and no problems at runtime - the key will just be what the key is. it might break apps that rely on this conversion, but I highly doubt it since it's not mentioned.

Reply:
That’s a really good point @TkDodo! I looked at the code again and managed to find a solution in which we don’t don’™ use typecasting, but all tests still pass.

`QueryFunctionContext` gets the original `TQueryKey`, while `FetchContext` gets the one with `ensureArray` around it.

Reply:
To be honest, all test still pass even if I remove `ensureArray` in `FetchContext`. Not sure if this is good or bad 😁

Reply:
as far as i can see, `context` is mainly used for `onFetch` of the `infiniteQueryBehavior`. But the type of the queryKey is still `TQueyKey`, which could also be a string on type level. so if you want to use `context.queryKey`, you cannot rely on it being an Array, which is why I think we don't need _ensureArray_ at all, too :)

`context.queryKey` is also only used to pass it to the `queryFn` in `infiniteQueryBehavior`, which is the same use-case as passing it here, so we would have the same runtime problem with infinite queries like I described above.

---

So my vote would go towards removing the `ensureArray` call here completely, but I'd like @tannerlinsley or @boschni to confirm before we do this please :)

Reply:
Interesting – I did not know about `infiniteQueryBehavior`. Makes sense! I removed `ensureArray` in 415e4a41d2b74e44c01300bfe4b7a3d86a3d7d35 to demonstrate passing tests. Happy to revert if that was not needed!

Reply:
@tannerlinsley or @boschni WDYT?

Reply:
Think it would indeed be better to remove it as this will make sure the query key type is consistent throughout the application.

Reply:
it means we'd have a small runtime change as well, but I think it would still qualify as a patch release. transforming the query key to an array when passing it to the queryFn via queryContext was not documented, so I think this is good 👍 .

Reply:
Thanks for your replies folks! Please let me know if you’d like me to make any further edits 🙂

Reply:
@boschni are you fine with a patch release?

Reply:
Yes I'm fine with a patch release. The current types should already prevent users from using it in such a way.